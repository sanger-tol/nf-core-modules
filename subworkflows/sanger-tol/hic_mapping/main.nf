include { BWAMEM2_INDEX              } from '../../../modules/nf-core/bwamem2/index/main'
include { HICCRAMALIGN_CHUNKS        } from '../../../modules/sanger-tol/hiccramalign/chunks'
include { HICCRAMALIGN_BWAMEM2ALIGN  } from '../../../modules/sanger-tol/hiccramalign/bwamem2align'
include { HICCRAMALIGN_MINIMAP2ALIGN } from '../../../modules/sanger-tol/hiccramalign/minimap2align'
include { MINIMAP2_INDEX             } from '../../../modules/nf-core/minimap2/index/main'
include { SAMTOOLS_FAIDX             } from '../../../modules/nf-core/samtools/faidx/main'
include { SAMTOOLS_INDEX             } from '../../../modules/nf-core/samtools/index/main'
include { SAMTOOLS_MARKDUP           } from '../../../modules/nf-core/samtools/markdup/main'
include { SAMTOOLS_MERGE             } from '../../../modules/nf-core/samtools/merge/main'

workflow HIC_MAPPING {

    take:
    ch_assemblies        // Channel [meta, assembly]
    ch_hic_cram          // Channel [meta, cram] OR [meta, [cram1, cram2, ..., cram_n]]
    val_aligner          // string: [either "bwamem2" or "minimap2"]
    val_cram_chunk_size  // integer: Number of CRAM slices per chunk for mapping
    val_mark_duplicates  // boolean: Mark duplicates on the BAM?

    main:
    ch_versions = Channel.empty()

    //
    // Logic: check if CRAM files are accompanied by an index
    //        Get indexes, and index those that aren't
    //
    ch_hic_cram_raw = ch_hic_cram
        | transpose()
        | branch { meta, cram ->
            def cram_file = file(cram, checkIfExists: true)
            def index = cram + ".crai"
            have_index: file(index).exists()
                return [ meta, cram_file, file(index, checkIfExists: true) ]
            no_index: true
                return [ meta, cram_file ]
        }

    //
    // Module: Index CRAM files without indexes
    //
    SAMTOOLS_INDEX(ch_hic_cram_raw.no_index)
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions)

    ch_hic_cram_indexed = ch_hic_cram_raw.have_index
        | mix(
            ch_hic_cram.no_index.join(SAMTOOLS_INDEX.out.crai)
        )

    //
    // Module: Process the cram index files to determine how many
    //         chunks to split into for mapping
    //
    HICCRAMALIGN_CHUNKS(
        ch_hic_cram_indexed,
        val_cram_chunk_size
    )
    ch_versions = ch_versions.mix(HICCRAMALIGN_CHUNKS.out.versions)

    //
    // Logic: Count the total number of cram chunks for downstream grouping
    //
    ch_n_cram_chunks = HICCRAMALIGN_CHUNKS.out.cram_slices
        | map { _meta, _cram, _crai, chunkn, _slices -> chunkn }
        | collect
        | map { chunkns -> chunkns.size() }

    //
    // Logic: Begin alignment - fork depending on specified aligner
    //
    if(val_aligner == "bwamem2") {
        //
        // Module: Create bwa-mem2 index for assembly
        //
        BWAMEM2_INDEX(ch_assemblies)
        ch_versions = ch_versions.mix(BWAMEM2_INDEX.out.versions)

        ch_assemblies_with_reference = ch_assemblies
            | combine(BWAMEM2_INDEX.out.index, by: 0)

        ch_cram_chunks = HICCRAMALIGN_CHUNKS.out.cram_slices
            | transpose()
            | combine(ch_assemblies_with_reference)
            | map { _meta, cram, crai, chunkn, slices, meta_assembly, index, assembly ->
                [ meta_assembly, cram, crai, chunkn, slices, index, assembly ]
            }

        HICCRAMALIGN_BWAMEM2ALIGN(ch_cram_chunks)
        ch_versions = ch_versions.mix(HICCRAMALIGN_BWAMEM2ALIGN.out.versions)

        ch_mapped_bams = HICCRAMALIGN_BWAMEM2ALIGN.out.bam
    } else if(val_aligner == "minimap2") {
        //
        // MODULE: generate minimap2 mmi file
        //
        MINIMAP2_INDEX(ch_assemblies)
        ch_versions = ch_versions.mix(MINIMAP2_INDEX.out.versions)

        ch_cram_chunks = HICCRAMALIGN_CHUNKS.out.cram_slices
            | transpose()
            | combine(MINIMAP2_INDEX.out.index)
            | map { _meta, cram, crai, chunkn, slices, meta_assembly, index ->
                [ meta_assembly, cram, crai, chunkn, slices, index ]
            }

        HICCRAMALIGN_MINIMAP2ALIGN(ch_cram_chunks)
        ch_versions = ch_versions.mix(HICCRAMALIGN_MINIMAP2ALIGN.out.versions)

        ch_mapped_bams = HICCRAMALIGN_MINIMAP2ALIGN.out.bam
    } else {
        log.error("Unsupported aligner: ${val_aligner}")
    }

    //
    // Logic: Index assembly fastas
    //
    SAMTOOLS_FAIDX(
        ch_assemblies, // reference
        [[:],[]],   // fai
        false       // get sizes
    )
    ch_versions = ch_versions.mix(SAMTOOLS_FAIDX.out.versions)

    //
    // Logic: Prepare input for merging bams.
    //        We use the ch_n_cram_chunks to set a groupKey so that
    //        we emit groups downstream ASAP once all bams have been made
    //
    ch_samtools_merge_input = ch_mapped_bams
        | combine(ch_n_cram_chunks)
        | map { meta, bam, n_chunks ->
            def key = groupKey(meta, n_chunks)
            [key, bam]
        }
        | groupTuple()
        | map { key, bam -> [key.target, bam] } // Get meta back out of groupKey
        | combine(ch_assemblies, by: 0)
        | combine(SAMTOOLS_FAIDX.out.fai, by: 0)
        | combine(SAMTOOLS_FAIDX.out.gzi.ifEmpty([[], []]), by: 0)
        | multiMap { meta, bams, assembly, fai, gzi ->
            bam:   [meta, bams]
            fasta: [meta, assembly]
            fai:   [meta, fai]
            gzi:   [meta, gzi]
        }

    //
    // Module: Merge position-sorted bam files
    //
    SAMTOOLS_MERGE(
        ch_samtools_merge_input.bam,
        ch_samtools_merge_input.fasta,
        ch_samtools_merge_input.fai,
        ch_samtools_merge_input.gzi,
    )
    ch_versions = ch_versions.mix(SAMTOOLS_MERGE.out.versions)

    //
    // Module: Mark duplicates on the merged bam
    //
    ch_samtools_markdup_input = SAMTOOLS_MERGE.out.bam
        | filter { val_mark_duplicates }
        | combine(ch_assemblies, by: 0)
        | multiMap { meta, bam, assembly ->
            bam:      [meta, bam]
            assembly: [meta, assembly]
        }

    SAMTOOLS_MARKDUP(
        ch_samtools_markdup_input.bam,
        ch_samtools_markdup_input.assembly
    )
    ch_versions = ch_versions.mix(SAMTOOLS_MARKDUP.out.versions)

    emit:
    bam      = val_mark_duplicates ? SAMTOOLS_MARKDUP.out.bam : SAMTOOLS_MERGE.out.bam
    versions = ch_versions
}
