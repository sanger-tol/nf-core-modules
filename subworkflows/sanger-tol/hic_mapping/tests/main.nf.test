nextflow_workflow {

    name "Test Subworkflow HIC_MAPPING"
    script "../main.nf"
    workflow "HIC_MAPPING"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_sangertol"
    tag "subworkflows/hic_mapping"
    tag "cramalign/gencramchunks"
    tag "cramalign/bwamem2alignhic"
    tag "cramalign/minimap2alignhic"
    tag "subworkflows/../../modules/nf-core/samtools/markdup"
    tag "subworkflows/../../modules/nf-core/minimap2/index"
    tag "subworkflows/../../modules/nf-core/samtools/merge"
    tag "subworkflows/../../modules/nf-core/samtools/index"
    tag "subworkflows/../../modules/nf-core/samtools/faidx"
    tag "subworkflows/../../modules/nf-core/bwamem2/index"

    setup {
        nfcoreInitialise("${launchDir}/library/")
        nfcoreInstall(
            "${launchDir}/library/",
            [
                "bwamem2/index",
                "minimap2/index",
                "samtools/index",
                "samtools/merge",
                "samtools/markdup",
                "samtools/faidx"
            ]
        )
        nfcoreLink("${launchDir}/library/", "${baseDir}/modules/")
    }

    test("meles meles - bwamem2") {

        when {

            params {
                samtools_cat_args = ""
                samtools_fastq_args = "-F0xB00 -nt"
                bwamem2_mem_args = "-5SPp"
                samtools_fixmate_args = "-mpu"
                samtools_view_args = "-q 0 -F 0x904"
                samtools_sort_args = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = "bwamem2"
                input[3] = 1
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - bwamem2 - multiple samples") {

        when {

            params {
                samtools_cat_args = ""
                samtools_fastq_args = "-F0xB00 -nt"
                bwamem2_mem_args = "-5SPp"
                samtools_fixmate_args = "-mpu"
                samtools_view_args = "-q 0 -F 0x904"
                samtools_sort_args = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test2' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true)
                    ],
                    [
                        [ id:'test2' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = "bwamem2"
                input[3] = 1
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 2 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - bwamem2 - multiple samples - meta mismatch") {

        when {

            params {
                samtools_cat_args = ""
                samtools_fastq_args = "-F0xB00 -nt"
                bwamem2_mem_args = "-5SPp"
                samtools_fixmate_args = "-mpu"
                samtools_view_args = "-q 0 -F 0x904"
                samtools_sort_args = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ],
                    [
                        [ id:'test2' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true)
                    ],
                    [
                        [ id:'test3' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = "bwamem2"
                input[3] = 1
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

    test("meles meles - minimap2") {

        when {
            params {
                samtools_cat_args     = ""
                samtools_fastq_args   = "-F0xB00 -nt"
                minimap2_args         = "-ax sr"
                samtools_fixmate_args = "-mpu"
                samtools_view_args    = "-q 0 -F 0x904"
                samtools_sort_args    = "--write-index -l1"
            }

            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ], // meta map
                        file(params.modules_testdata_base_path + 'Meles_meles/assembly/release/mMelMel3.1_paternal_haplotype/GCA_922984935.2.subset.fasta.gz', checkIfExists: true)
                    ]
                )
                input[1] = Channel.of(
                    [
                        [ id:'test' ],
                        file(params.modules_testdata_base_path + 'Meles_meles/genomic_data/mMelMel3/hic-arima2/35528_2%231_subset.cram', checkIfExists: true)
                    ]
                )
                input[2] = "minimap2"
                input[3] = 1
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert workflow.out.bam.size() == 1 },
                {
                    assert snapshot(
                        workflow.out.bam.collect { bam(it.get(1)).getStatistics() }.sort(),
                        workflow.out.versions
                    ).match()
                }
            )
        }

        cleanup {
            nfcoreUnlink("${launchDir}/library/", "${baseDir}/modules/nf-core")
        }
    }

}
