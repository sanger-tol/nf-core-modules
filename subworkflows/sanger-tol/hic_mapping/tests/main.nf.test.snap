{
    "meles meles - bwamem2 - multiple samples": {
        "content": [
            [
                {
                    "maxReadLength": 151,
                    "minReadLength": 151,
                    "meanReadLength": 151,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 7,
                    "readCount": 3802,
                    "duplicateReadCount": 54,
                    "sorted": true
                },
                {
                    "maxReadLength": 151,
                    "minReadLength": 151,
                    "meanReadLength": 151,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 7,
                    "readCount": 3802,
                    "duplicateReadCount": 54,
                    "sorted": true
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,6da1ff684dd357b0be447eff01112c09",
                "versions.yml:md5,6da1ff684dd357b0be447eff01112c09",
                "versions.yml:md5,804d2da0c4ff5dfe1ea5b14f0c3994a2",
                "versions.yml:md5,804d2da0c4ff5dfe1ea5b14f0c3994a2",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,e938bc5e520e0423caa5e6ed7908cb3d",
                "versions.yml:md5,e938bc5e520e0423caa5e6ed7908cb3d",
                "versions.yml:md5,fdcbb8a03dee4c8ba03fd83bbcd10f19",
                "versions.yml:md5,fdcbb8a03dee4c8ba03fd83bbcd10f19"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-27T09:32:34.01216"
    },
    "meles meles - bwamem2": {
        "content": [
            [
                {
                    "maxReadLength": 151,
                    "minReadLength": 151,
                    "meanReadLength": 151,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 7,
                    "readCount": 3802,
                    "duplicateReadCount": 54,
                    "sorted": true
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,6da1ff684dd357b0be447eff01112c09",
                "versions.yml:md5,804d2da0c4ff5dfe1ea5b14f0c3994a2",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,e938bc5e520e0423caa5e6ed7908cb3d",
                "versions.yml:md5,fdcbb8a03dee4c8ba03fd83bbcd10f19"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-27T09:31:41.522091"
    },
    "meles meles - minimap2": {
        "content": [
            [
                {
                    "maxReadLength": 151,
                    "minReadLength": 25,
                    "meanReadLength": 147,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 9,
                    "readCount": 1856,
                    "duplicateReadCount": 33,
                    "sorted": true
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,6da1ff684dd357b0be447eff01112c09",
                "versions.yml:md5,897099b3ba1f2612a98b07f805a89d42",
                "versions.yml:md5,897099b3ba1f2612a98b07f805a89d42",
                "versions.yml:md5,e938bc5e520e0423caa5e6ed7908cb3d",
                "versions.yml:md5,fd15db333caf07479a96afdbe9a149e0",
                "versions.yml:md5,fdcbb8a03dee4c8ba03fd83bbcd10f19"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-27T09:34:05.763994"
    },
    "meles meles - bwamem2 - multiple samples - meta mismatch": {
        "content": [
            [
                {
                    "maxReadLength": 151,
                    "minReadLength": 151,
                    "meanReadLength": 151,
                    "maxQuality": 60,
                    "minQuality": 0,
                    "meanQuality": 7,
                    "readCount": 3802,
                    "duplicateReadCount": 54,
                    "sorted": true
                }
            ],
            [
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,0ba5a443eb8721c533ff4d72982054f9",
                "versions.yml:md5,6da1ff684dd357b0be447eff01112c09",
                "versions.yml:md5,804d2da0c4ff5dfe1ea5b14f0c3994a2",
                "versions.yml:md5,804d2da0c4ff5dfe1ea5b14f0c3994a2",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,bbc059343be8756eb41864dceae346d0",
                "versions.yml:md5,e938bc5e520e0423caa5e6ed7908cb3d",
                "versions.yml:md5,e938bc5e520e0423caa5e6ed7908cb3d",
                "versions.yml:md5,fdcbb8a03dee4c8ba03fd83bbcd10f19"
            ]
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-10-27T09:33:18.58961"
    }
}